MetaMask/metamask-mobile
# feat: Eth snap keyring

## **Description**

1. What is the reason for the change?
- This change adds lays  the foundation for third party snap accounts by implementing/integrating the [eth-snap-keyring](https://github.com/MetaMask/eth-snap-keyring) into mobile flask
- this is flask only feature
- this change is needed to unlock the future of account management beyond native EOA accounts. This will enable third party MPC accounts such as [Capsule](https://usecapsule.com/). This also paves the way for multi chain accounts such as bitcoin.
- This work has already been implemented and launched on the extension in production. We are finally going to bring these feature to mobile.
2. What is the improvement/solution?
- implement the SnapKeyring builder. This implementation mirrors the [one on extension](https://github.com/MetaMask/metamask-extension/tree/develop/app/scripts/lib/snap-keyring).
- Implement and wire up the necessary snaps methods (getAllowedKeyringMethods, handleSnapRpcRequest, getSnapKeyring, hasPermission). The methods need to be implemented in the `Engine` as well as the `snapMethodMiddlewareBuilder`
- add `'endowment:keyring': 'endowment:keyring'` to the set of available snap permissions
- implement `keyringSnapPermissionsBuilder` which derives the permissions for a specific origin.
- include the new SnapKeyring in the [additionalKeyring](https://github.com/MetaMask/metamask-mobile/pull/10829/files#diff-1a656ef42f595bb9c6c1251b29808fd70d31da7a3e1a6e9be85574ad866dc1b0R796-R803) section of the keyring controller. This is where we add non "native" keyrings such as ledger and qr hardware wallets. The snapkeyring will now be included in this list `IF THE BUILD TYPE IS FLASK`.
- I had to make some controllers properties on the Engine class for easier access across the file. This matches more closely what the extension is doing in the `metamask-controllers.js` file. 


### This PR DOES NOT
- Allow users to confirm or deny snap account addition/removal. For now we will accept all incoming snap account additions/removals. Those confirmations will be handled in these following changes. https://github.com/MetaMask/accounts-planning/issues/152 and https://github.com/MetaMask/accounts-planning/issues/151
- This PR does not support custom for account snaps
- This PR does not support asynchronous approvals 
- This PR does not implement the improved account list U.I for snap accounts. For now the snap accounts in the account list will look exactly like the rest of the accounts.
- This PR does not support granular permissions required in production since there seems to be come issues with the mobile snaps permissions in general. see https://github.com/MetaMask/metamask-mobile/issues/11029. For now we will eagerly grant keyring permissions to dapps since this is just a flask feature and is not doing into production.
- Restrict wallet usage based on account permissions. Even if an account does not have signing permissions these flows will be available in the U.I and likely break.


## **Related issues**

Fixes: https://github.com/MetaMask/accounts-planning/issues/149

## **Manual testing steps**

1. run yarn setup on this branch
2. navigate to .js.env and set `export METAMASK_BUILD_TYPE=flask`
5. `source .js.env`
6. `yarn watch:clean`
7. create/import a wallet
8. navigate to settings/snaps
9. there should be only one pre installed snap
10. open the browser
11. navigate to `https://metamask.github.io/snap-simple-keyring/latest/`
12. click connect and accept all of the permissions
13. you should now see this snap in the settings/snaps list
14. click create account on the snap simple keyring
15. the call should succeed and return the account info as a json obkect
16. navigate to the home page and open the account list, the selected account should be a new account named `Snap Account 1`
17. repeat step 15 as many times as you want, each time the new account should be added
18. clicking list accounts on the SSK dapp should give you the entire list of accounts
19. extract an id value from the listed accounts and paste it into the Remove accounts field and click.
20. the call should succeed and return null
21. navigate back to the home page and open the account list, that account should no longer be there and another account should be the selected account
22. 

## **Screenshots/Recordings**



### **Before**

N/A

### **After**


https://github.com/user-attachments/assets/79b0a22d-2bf3-4124-883d-82b91ab0121f



## **Pre-merge author checklist**


## **Pre-merge reviewer checklist**
